### PR TITLE
PMS-9, PMS-13 Check Transactions & Filter and sorting items

### PR DESCRIPTION
**Pull request by erik-lance**
Screenshot_284

Now has a route at /transactions which will show you all transactions.

Has a transaction schema. Also contains the "name" of the item secretly. Only used for searching to simplify search times.

Can filter based on
- name
- description (especially the code). 

Every new item added creates a new transaction.

### CODE REVIEW

File: controllers/transactionController.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    // The transaction page
+    transactions: function (req, res) {
+        res.render("transactions", {
+            title: "Transactions",
+            styles: ["index.css", "w2ui-overrides.css"],
+            scripts: ["transaction.js"],
+        });
+    },
+
+    addTransaction: async function (req, res) {
```
Comment (aar0npal) [inline controllers/transactionController.js:16]: Add `try...catch`
In `catch`, return `500` with `message: Server error`

File: controllers/transactionController.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+        console.log('Added transaction: ')
+        console.log(transItem)
+
+        db.insertOne(Transaction, transItem, function (data) {
+            res.send(data);
+        });
+    },
+
+    getTransactions: function (req, res) {
```
Comment (aar0npal) [inline controllers/transactionController.js:44]: Add `try...catch`
In `catch`, return `500` with `message: Server error`

File: controllers/transactionController.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        db.findMany(Transaction, {}, null, function (data) {
+            res.status(200).json(data);
+        });
+    },
+
+    // UNTESTED
+    getTransaction: function (req, res) {
```
Comment (aar0npal) [inline controllers/transactionController.js]: Add `try...catch`
In `catch`, return `500` with `message: Server error`

File: controllers/transactionController.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        db.findOne(Transaction, {description: {$regex:req.query.code, $options: 'i'}}, {}, async function(data) {
+
+            console.log(req.query)
+            res.status(200).json(await data);
+        })
+
+    },
+
+    searchTransactions: function (req, res) {
```
Comment (aar0npal) [inline controllers/transactionController.js:110]: Add `try...catch`
In `catch`, return `500` with `message: Server error`

File: controllers/transactionController.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
+//Controller for items
+import Transaction from "../model/schemas/Transaction.js";
+import db from "../model/db.js";
+
+const transactionController = {
+    // The transaction page
+    transactions: function (req, res) {
+        res.render("transactions", {
+            title: "Transactions",
+            styles: ["index.css", "w2ui-overrides.css"],
```
Comment (aar0npal) [inline controllers/transactionController.js]: Move to `pages` and `general` folder

**Review by aar0npal — CHANGES_REQUESTED** [review]

File: controllers/transactionController.js (hunk 1 above)
Comment (GelsonSze) [reply controllers/transactionController.js:16]: see [cd1c739](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/cd1c739754a0c62c4b2346b66ee0e497f871eb46)

File: controllers/transactionController.js (hunk 2 above)
Comment (GelsonSze) [reply controllers/transactionController.js:44]: see [cd1c739](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/cd1c739754a0c62c4b2346b66ee0e497f871eb46)

File: controllers/transactionController.js (hunk 3 above)
Comment (GelsonSze) [reply controllers/transactionController.js]: see [cd1c739](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/cd1c739754a0c62c4b2346b66ee0e497f871eb46)

File: controllers/transactionController.js (hunk 4 above)
Comment (GelsonSze) [reply controllers/transactionController.js:110]: see [cd1c739](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/cd1c739754a0c62c4b2346b66ee0e497f871eb46)

File: controllers/transactionController.js (hunk 5 above)
Comment (erik-lance) [reply controllers/transactionController.js]: see 0489b59349b88012dbf6d15c8f61abe8578eab2e

**Review by aar0npal — APPROVED** [review]
LGTM! Good work.
Note: pa-ayos nalang ng CSS for table filter (specifically `#table-filter-...` should be class instead)

File: public/css/pages/index.css [hunk 6]
```diff
@@ -127,7 +127,7 @@ main {
     cursor: pointer;
 }
 
-#table-filter-apply {
+#table-filter-apply, #table-filter-clear {
```
Comment (aar0npal) [inline public/css/pages/index.css]: Referring to the note above:
Use class `.table-filter` instead of id for general styling.

**Review by SeanUmpad — CHANGES_REQUESTED** [review]
**Checked:**
Sorters - Working except the date sorter
Filters - Working except search filter has a bug

**Test Cases:** https://docs.google.com/spreadsheets/d/1-uWENiTCqkX6zCxiP422kY14om7W3PoVnO5XfAQxOH8

**Errors:**
Test ID 033: when sorting by date the time is sorted in descending order (should be ascending right?)
image

Test ID 043: search filter bugs messes up the description
**BEFORE**
image
**AFTER**
image

**Comment by SeanUmpad** [comment]
**Checked:**
Test ID 033: when sorting by date the time is sorted in descending order - FIXED
Test ID 043: search filter bugs messes up the description - FIXED

**Errors:**
Test ID 049: clicking apply filter doesnt refresh the table until you adjust the screen with f12/ctrl+-/etc
image

Test ID 050: clicking apply filter too fast consecutively duplicates the shown transactions
image

File: public/css/pages/index.css (hunk 6 above)
Comment (GelsonSze) [reply public/css/pages/index.css]: refer to [bade27d](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/bade27df50803c2d1b36867abf1f53d09f8dae21)

**Comment by SeanUmpad** [comment]
**Checked:**
Test ID 049: clicking apply filter doesnt refresh the table until you adjust the screen with f12/ctrl+-/etc
image - FIXED
Test ID 050: clicking apply filter too fast consecutively duplicates the shown transactions - FIXED

**No Errors**

**Review by SeanUmpad — APPROVED** [review]

File: controllers/transactionController.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
+//Controller for items
```
Comment (aar0npal) [inline controllers/transactionController.js:1]: Just for consistency with other files, rename `err` to `error` in catch.
Also, for `details: err`  -> must be `details: error.message`

File: public/js/item.js [hunk 8]
```diff
@@ -35,6 +36,121 @@ function Item(
...
+
+            window.location.href = "/item/" + code;
+        },
+    });
+
+    $.ajax({
+        url: `/searchTransactions=${"Type"}&${itemCode}`,
+        type: "GET",
+        processData: false,
+        contentType: false,
```
Comment (aar0npal) [inline public/js/item.js]: Refactor: 
- remove `headers`
- use `contentType` instead; `contentType: "application/json; charset=utf-8"`

File: public/js/item.js [hunk 9]
```diff
@@ -35,6 +36,121 @@ function Item(
...
+ */
+function pushTransaction(trans) {
+    var dfd = $.Deferred();
+
+    $.ajax({
+        url: `/getItemById=${trans.description}`,
+        type: "GET",
+        processData: false,
+        contentType: false,
+        headers: {
```
Comment (aar0npal) [inline public/js/item.js]: Refactor: 
- remove `headers`
- use `contentType` instead; `contentType: "application/json; charset=utf-8"`

File: public/js/index.js [hunk 10]
```diff
@@ -17,27 +17,22 @@ function getAllItems(refreshGrid = false) {
                 "Content-Type": "application/json",
```
Comment (aar0npal) [inline public/js/index.js]: Refactor: 
- remove `headers`
- use `contentType` instead; `contentType: "application/json; charset=utf-8"`

File: public/js/index.js [hunk 11]
```diff
@@ -74,6 +69,87 @@ function item(
...
+    var check = $("#filter-search").val().toLowerCase();
+
+    /*Process gets all items given a specific condition, which is if the item has the following category. The ==0 condition
+    is only for the instances where the category was not changed.*/
+    $.ajax({
+        url: "/getItems",
+        type: "GET",
+        processData: false,
+        contentType: false,
+        headers: {
```
Comment (aar0npal) [inline public/js/index.js]: Refactor: 
- remove `headers`
- use `contentType` instead; `contentType: "application/json; charset=utf-8"`

File: public/js/item.js [hunk 12]
```diff
@@ -46,6 +162,7 @@ function getItem() {
         type: "GET",
         processData: false,
         contentType: false,
+        data: { code: item_code },
         headers: {
```
Comment (aar0npal) [inline public/js/item.js]: Refactor: 
- remove `headers`
- use `contentType` instead; `contentType: "application/json; charset=utf-8"`

File: public/js/transaction.js [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+ * render it in the grid.
+ * @param  {boolean} [refreshGrid=false] - If true, render the data in the grid.
+ */
+function getAllTransactions(refreshGrid = false) {
+    $.ajax({
+        url: "/getTransactions",
+        type: "GET",
+        processData: false,
+        contentType: false,
+        headers: {
```
Comment (aar0npal) [inline public/js/transaction.js]: Refactor: 
- remove `headers`
- use `contentType` instead; `contentType: "application/json; charset=utf-8"`

File: public/js/transaction.js [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+ */
+function pushTransaction(trans) {
+    var dfd = $.Deferred();
+
+    $.ajax({
+        url: `/getItemById=${trans.description}`,
+        type: "GET",
+        processData: false,
+        contentType: false,
+        headers: {
```
Comment (aar0npal) [inline public/js/transaction.js]: Refactor: 
- remove `headers`
- use `contentType` instead; `contentType: "application/json; charset=utf-8"`

File: public/js/transaction.js [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+    if (searchBar == "empty" && typeBar == "Type") {
+        getAllTransactions(true);
+    } else {
+        $.ajax({
+            url: `/searchTransactions=${typeBar}&${searchBar}`,
+            type: "GET",
+            processData: false,
+            contentType: false,
+            headers: { "Content-Type": "application/json" },
```
Comment (aar0npal) [inline public/js/transaction.js]: Refactor: 
- remove `headers`
- use `contentType` instead; `contentType: "application/json; charset=utf-8"`

File: public/js/utils.js [hunk 16]
```diff
@@ -223,18 +223,35 @@ function formatDate(d) {
...
+ */
+function formatDate(d) {
+    //include 0 if the year is less than 10
+    const year = d.getFullYear() < 10 ? "0" + d.getFullYear() : d.getFullYear();
+    //include 0 if the month is less than 10
+    const month = d.getMonth()+1 < 10 ? `0${d.getMonth()+1}` : d.getMonth()+1;
+    //include 0 if the day is less than 10
+    const day = d.getDate() < 10 ? `0${d.getDate()}` : d.getDate();
+
+    return `${year}/${month}/${day} ${formatAMPM(d)}`;
```
Comment (aar0npal) [inline public/js/utils.js]: Must be `${month+1}`

Even though the fix is in PMS-4, lets apply this para maayos na pag merge sa development.

File: controllers/itemController.js [hunk 17]
```diff
@@ -2,13 +2,13 @@
 import Item from "../model/schemas/Item.js";
```
Comment (aar0npal) [inline controllers/itemController.js:2]: Just for consistency with other files, rename `err` to `error` in catch.
Also, for `details: err`  -> must be `details: error.message`

File: sample/testController.js [hunk 18]
```diff
@@ -1,5 +1,6 @@
 //Controller for sample data
```
Comment (aar0npal) [inline sample/testController.js:1]: Just for consistency with other files, rename `err` to `error` in catch.
Also, for `details: err`  -> must be `details: error.message`

File: views/layouts/main.hbs [hunk 19]
```diff
@@ -28,7 +28,7 @@
         {{#each scripts}}
         <script src="/js/{{this}}"></script>
         {{/each}}
-        <title>{{title}} - TitaFindsXG5</title>
+        <title>{{title}} - TitaFinds</title>
```
Comment (aar0npal) [inline views/layouts/main.hbs]: Must be "TitaFinds Inventory"

**Review by aar0npal — CHANGES_REQUESTED** [review]
@GelsonSze speedrun kaya?

File: controllers/transactionController.js (hunk 7 above)
Comment (GelsonSze) [reply controllers/transactionController.js:1]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

File: public/js/item.js (hunk 8 above)
Comment (GelsonSze) [reply public/js/item.js]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

File: public/js/index.js (hunk 10 above)
Comment (GelsonSze) [reply public/js/index.js]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

File: public/js/index.js (hunk 11 above)
Comment (GelsonSze) [reply public/js/index.js]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

File: public/js/item.js (hunk 9 above)
Comment (GelsonSze) [reply public/js/item.js]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

File: public/js/item.js (hunk 12 above)
Comment (GelsonSze) [reply public/js/item.js]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

File: public/js/transaction.js (hunk 13 above)
Comment (GelsonSze) [reply public/js/transaction.js]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

File: public/js/transaction.js (hunk 14 above)
Comment (GelsonSze) [reply public/js/transaction.js]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

File: public/js/utils.js (hunk 16 above)
Comment (GelsonSze) [reply public/js/utils.js]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

File: public/js/transaction.js (hunk 15 above)
Comment (GelsonSze) [reply public/js/transaction.js]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

File: sample/testController.js (hunk 18 above)
Comment (GelsonSze) [reply sample/testController.js:1]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

File: views/layouts/main.hbs (hunk 19 above)
Comment (GelsonSze) [reply views/layouts/main.hbs]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

File: controllers/itemController.js (hunk 17 above)
Comment (GelsonSze) [reply controllers/itemController.js:2]: see [ff9f289](https://github.com/erik-lance/TitafindsXG5/pull/11/commits/ff9f28905aaea1d93ce0ee0f4de2ade76e361ff1)

**Review by aar0npal — APPROVED** [review]

**Pull request merged**